### PR TITLE
cpp version mismatch diagnostics + tests

### PR DESCRIPTION
**Pull request by jgstarIntel**
**Summary**
Adding auto-generated C++ version mismatch diagnostics to the integration header generation files to allow version mismatch detection for separate compilation flows.

**Detail**
An issue occurs when the C++ version used for the device compilation equals C++14 and the version used during the host compilation equals  C++11. For example, when the representation of an object in one C++ version by a compiler is different from that of another C++ version for the same object.
 
With this fix, the two versions (device C++ version and host C++ version) will be captured in the auto-generated integration header. When the integration header is run during the host compilation, the versions will be compared and a diagnostic can be reported.
 
This change set specifically covers cases where: C++ version (std=C++11 or less) for host compilation cannot be matched with C++ version (std=C++14 or greater) for device compilation.

### CODE REVIEW

File: sycl/test/separate-compile/cpp_version_mismatch_test_2.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+//==----------- cpp_version_mismatch_test_2.cpp - SYCL separate compilation cpp version mismatch test -----------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// -----------------------------------------------------------------------------
+#include <CL/sycl.hpp>
+#include <iostream>
```
Comment (rolandschulz) [inline sycl/test/separate-compile/cpp_version_mismatch_test_2.cpp:23]: You don't need an kernel doing something for this test, do you? An empty kernel is sufficient.

```suggestion

namespace sycl = cl::sycl;

int main() {
    sycl::queue().submit([&](sycl::handler& cgh) {
        cgh.parallel_for<class task>(sycl::nd_range<1>(16,16),
            [=](auto item) {
            });});
}

BTW: Why don't we write out an empty integration header if `-fsycl-int-header` is provided but there is no kernel in the source file? I can see this causing an issue if users set up their build system but some files don't (yet) have a kernel. Than the required include file is missing.
```

File: sycl/test/separate-compile/cpp_version_mismatch_test_2.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+//
+// >> ---- host compilation: no cpp version mismatch
+// RUN: %clangxx -std=c++14 -include sycl_ihdr_a.h -c %s -I %sycl_include
+//
+// >> ---- code generation checks
+// RUN: FileCheck -input-file=sycl_ihdr_a.h %s
+// CHECK: #define STD_CPP_VERSION 
+// CHECK-NEXT: #if __cplusplus != STD_CPP_VERSION
+// CHECK-NEXT: #error "C++ version for host compilation does not match C++ version used for device compilation"
+// CHECK-NEXT: #endif
```
Comment (Fznamznon) [inline sycl/test/separate-compile/cpp_version_mismatch_test_2.cpp:12]: Please do integration header testing in FE tests. Such tests are placed in clang/test/CodeGenSYCL .

File: sycl/test/separate-compile/cpp_version_mismatch_test_1.cpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
+// >> ---- device compilation
+// RUN: %clangxx -std=c++14 -fsycl-device-only -Xclang -fsycl-int-header=sycl_ihdr_a.h %s -c -I %sycl_include
+//
+// >> ---- host compilation: cpp version mismatch
+// RUN: not %clangxx -std=c++11 -include sycl_ihdr_a.h -c %s -I %sycl_include 2>&1 | FileCheck %s 
+// CHECK: C++ version for host compilation does not match C++ version used for device compilation
```
Comment (Fznamznon) [inline sycl/test/separate-compile/cpp_version_mismatch_test_1.cpp:6]: I think you can make this test as FE test. Please move it to clang/test/SemaSYCL .

File: clang/lib/Sema/SemaSYCL.cpp [hunk 4]
```diff
@@ -2728,6 +2728,23 @@ static void emitKernelNameType(QualType T, ASTContext &Ctx, raw_ostream &OS,
   emitWithoutAnonNamespaces(OS, T.getCanonicalType().getAsString(TypePolicy));
 }
 
+int SYCLIntegrationHeader::getCppVersion() {
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:2731]: Doesn't clang already have similar stuff?
@erichkeane , don't you know?

File: clang/lib/Sema/SemaSYCL.cpp [hunk 5]
```diff
@@ -2737,6 +2754,16 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
 
   O << "\n";
 
+  int cpp_version = getCppVersion();
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:2757]: Please follow llvm coding guidelines.
See https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly

File: clang/lib/Sema/SemaSYCL.cpp (hunk 4 above)
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:2731]: I don't think we have exactly this function, but we DO similar logic for the __cpluplus macro in the preprocessor.

File: clang/include/clang/Sema/Sema.h [hunk 6]
```diff
@@ -322,6 +322,9 @@ class SYCLIntegrationHeader {
   SYCLIntegrationHeader(DiagnosticsEngine &Diag, bool UnnamedLambdaSupport,
                         Sema &S);
 
+  /// Get c++ version to support version mismatch diagnostics
```
Comment (erichkeane) [inline clang/include/clang/Sema/Sema.h:325]: I don't see a good reason for this to be a member function, it doesn't really seem to depend on the Sema or int-header.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 7]
```diff
@@ -2728,6 +2728,23 @@ static void emitKernelNameType(QualType T, ASTContext &Ctx, raw_ostream &OS,
   emitWithoutAnonNamespaces(OS, T.getCanonicalType().getAsString(TypePolicy));
 }
 
+int SYCLIntegrationHeader::getCppVersion() {
+  LangOptions LangOpts = S.getASTContext().getLangOpts();
+  if (LangOpts.CPlusPlus) {
+    if (LangOpts.CPlusPlus20)
+      return 201707L;
```
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:2735]: This value is wrong. C++20 is 202002L.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 8]
```diff
@@ -2728,6 +2728,23 @@ static void emitKernelNameType(QualType T, ASTContext &Ctx, raw_ostream &OS,
...
+    else if (LangOpts.CPlusPlus17)
+      return 201703L;
+    else if (LangOpts.CPlusPlus14)
+      return 201402L;
+    else if (LangOpts.CPlusPlus11)
+      return 201103L;
+    else
+      return 199711L;
+  }
+  return -1;
```
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:2745]: This seems odd and misguided.  Having this function allow !CPlusPlus at all seems like a mistake.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 9]
```diff
@@ -2728,6 +2728,23 @@ static void emitKernelNameType(QualType T, ASTContext &Ctx, raw_ostream &OS,
   emitWithoutAnonNamespaces(OS, T.getCanonicalType().getAsString(TypePolicy));
 }
 
+int SYCLIntegrationHeader::getCppVersion() {
+  LangOptions LangOpts = S.getASTContext().getLangOpts();
+  if (LangOpts.CPlusPlus) {
```
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:2733]: I'd likely replace this condition with an assert.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 10]
```diff
@@ -2737,6 +2754,16 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
 
   O << "\n";
 
+  int cpp_version = getCppVersion();
+  if (cpp_version != -1) {
+    O << "#define STD_CPP_VERSION ";
+    O << cpp_version << "\n";
+    O << "#if __cplusplus != STD_CPP_VERSION\n";
```
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:2761]: This is actually pretty risky when using other compilers.  MSVC for example used a 1997 date until just recently.

File: sycl/test/separate-compile/cpp_version_mismatch_test_2.cpp (hunk 2 above)
Comment (JoeyGIntel) [reply sycl/test/separate-compile/cpp_version_mismatch_test_2.cpp]: Capturing llvm-lit test transfer into a separate issue.

File: sycl/test/separate-compile/cpp_version_mismatch_test_1.cpp (hunk 3 above)
Comment (JoeyGIntel) [reply sycl/test/separate-compile/cpp_version_mismatch_test_1.cpp]: Capturing llvm-lit test folder transfer into a separate issue.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 10 above)
Comment (JoeyGIntel) [reply clang/lib/Sema/SemaSYCL.cpp]: Handling windows resolve and/or linux specifiers in a separate issue/ticket

File: clang/lib/Sema/SemaSYCL.cpp (hunk 5 above)
Comment (JoeyGIntel) [reply clang/lib/Sema/SemaSYCL.cpp]: Updated

File: clang/lib/Sema/SemaSYCL.cpp (hunk 8 above)
Comment (JoeyGIntel) [reply clang/lib/Sema/SemaSYCL.cpp]: Updated

File: clang/lib/Sema/SemaSYCL.cpp (hunk 7 above)
Comment (JoeyGIntel) [reply clang/lib/Sema/SemaSYCL.cpp]: Updated

File: clang/lib/Sema/SemaSYCL.cpp (hunk 9 above)
Comment (JoeyGIntel) [reply clang/lib/Sema/SemaSYCL.cpp]: Updated!

File: clang/include/clang/Sema/Sema.h (hunk 6 above)
Comment (JoeyGIntel) [reply clang/include/clang/Sema/Sema.h]: Updated and inlined in single instance where it's used

File: clang/lib/Sema/SemaSYCL.cpp [hunk 11]
```diff
@@ -2737,6 +2737,29 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
...
+  if (LangOpts.CPlusPlus20)
+    CppVersion = 202002L;
+  else if (LangOpts.CPlusPlus17)
+    CppVersion = 201703L;
+  else if (LangOpts.CPlusPlus14)
+    CppVersion = 201402L;
+  else if (LangOpts.CPlusPlus11)
+    CppVersion = 201103L;
+  else
+    CppVersion = 199711L;
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:2753]: I think that @erichkeane mentioned that it shouldn't be a **member** function. I think we can keep it as some function available in SemaSYCL .

File: sycl/test/separate-compile/cpp_version_mismatch_test_2.cpp (hunk 2 above)
Comment (Fznamznon) [reply sycl/test/separate-compile/cpp_version_mismatch_test_2.cpp]: What is the problem with llvm-lit?

**Review by Fznamznon — COMMENTED** [review]
> While I understand the motivation here, I think that there isn't a good way about this.
> 
> The __cplusplus macro is inconsistent between compilers that otherwise would be compatible (such as, we SHOULD be OK with a slightly older version of Clang where C++2a was 201911L).
> 
> Additionally, IS there a problem if the host is a slightly newer/older version?
> 
> Finally, this is going to be a maintenance headache (the rarity of is change, every 3 years, is going to result in us forgetting to update this) and massively error prone, and doesn't do what we need it to do.
> 
> I just think this is a bad idea.

I'm not going to approve this patch unless these concerns are resolved.

**Comment by romanovvlad** [comment]
@JoeyGIntel Could you please resolve concern mentioned above?

**Comment by bader** [comment]
@JoeyGIntel, please, fix pre-commit checks.

**Comment by rbegam** [comment]
@JoeyGIntel, sycl tests LGTM. Please address the review comments and fix the fails.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 12]
```diff
@@ -2728,13 +2726,45 @@ static void emitKernelNameType(QualType T, ASTContext &Ctx, raw_ostream &OS,
...
 
+  // Get c++ version to support version mismatch diagnostics
+  int CppVersion = getCppVersion(S);
+
+  // Generate C++ version mismatch diagnostics
+  O << "#define STD_CPP_VERSION ";
+  O << CppVersion << "\n";
+  O << "#if (__cplusplus <= 201112L) && (STD_CPP_VERSION >= 201401L) \n";
+  O << "#error \"C++ version (std=c++11 or less) for host compilation cannot "
+       "be matched with C++ version (std=c++14 or greater) for device "
```
Comment (elizabethandrews) [inline clang/lib/Sema/SemaSYCL.cpp:2765]: Can we change 'cannot be matched' to 'is incompatible'?

**Comment by elizabethandrews** [comment]
> While I understand the motivation here, I think that there isn't a good way about this.
> 
> The __cplusplus macro is inconsistent between compilers that otherwise would be compatible (such as, we SHOULD be OK with a slightly older version of Clang where C++2a was 201911L).
> 
> Additionally, IS there a problem if the host is a slightly newer/older version?
> 
> Finally, this is going to be a maintenance headache (the rarity of is change, every 3 years, is going to result in us forgetting to update this) and massively error prone, and doesn't do what we need it to do.
> 
> I just think this is a bad idea.

@erichkeane can you take a look at the latest patch. It has been modified to throw the error only when host version is C++11 (or less) and device version is C++14 (or greater)

**Comment by erichkeane** [comment]
> 
> 
> > While I understand the motivation here, I think that there isn't a good way about this.
> > The __cplusplus macro is inconsistent between compilers that otherwise would be compatible (such as, we SHOULD be OK with a slightly older version of Clang where C++2a was 201911L).
> > Additionally, IS there a problem if the host is a slightly newer/older version?
> > Finally, this is going to be a maintenance headache (the rarity of is change, every 3 years, is going to result in us forgetting to update this) and massively error prone, and doesn't do what we need it to do.
> > I just think this is a bad idea.
> 
> @erichkeane can you take a look at the latest patch. It has been modified to throw the error only when host version is C++11 (or less) and device version is C++14 (or greater)

This doesn't solve the problem actually.  The issue is that MSVC until recently (and on hosts we want to support), made this number always 1997.  It still supported a large amount of C++17 features at that point.

File: sycl/test/separate-compile/cpp_version_mismatch_test_2.cpp [hunk 13]
```diff
@@ -0,0 +1,38 @@
+// >> ---- device compilation
+// RUN: %clangxx -std=c++14 -fsycl-device-only -Xclang -fsycl-int-header=sycl_ihdr_a.h %s -c -I %sycl_include
+
+// >> ---- code generation checks
+// RUN: FileCheck -input-file=sycl_ihdr_a.h %s
+// CHECK: #define STD_CPP_VERSION
```
Comment (erichkeane) [inline sycl/test/separate-compile/cpp_version_mismatch_test_2.cpp:6]: Is there a good reason to extract this into a macro instead of just putting it right into the conditional?

File: sycl/test/separate-compile/cpp_version_mismatch_test_2.cpp [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+// CHECK-NEXT: #if (__cplusplus <= 201112L) && (STD_CPP_VERSION >= 201401L)
+// CHECK-NEXT: #error "C++ version (std=c++11 or less) for host compilation
```
Comment (erichkeane) [inline sycl/test/separate-compile/cpp_version_mismatch_test_2.cpp:8]: The wording needs work, but this is perhaps less harmful if it is a pragma-warning (assuming this isn't included as a system header).

File: sycl/test/separate-compile/cpp_version_mismatch_test_2.cpp (hunk 13 above)
Comment (elizabethandrews) [reply sycl/test/separate-compile/cpp_version_mismatch_test_2.cpp:6]: You mean just call `getCppVersion(S)` in conditional? I don't see why that can't be done. @JoeyGIntel

Comment (erichkeane) [reply sycl/test/separate-compile/cpp_version_mismatch_test_2.cpp:6]: Actually, we ALREADY know if STD_CPP_VERSION >= 201401L, so emitting it int-header is foolish.  We should just omit this enitre set of code if that isn't the case .

Comment (erichkeane) [reply sycl/test/separate-compile/cpp_version_mismatch_test_2.cpp:6]: AND, it simplifies the patch a ton, you don't even NEED getCppVersion anymore, since you can just do:

if (LangOpts.CPlusPlus14) {
  os << "#if (__cplusplus <=201112L)\n";
  os << "    #error ...\n";
|

Comment (premanandrao) [reply sycl/test/separate-compile/cpp_version_mismatch_test_2.cpp:6]: I like this idea a lot!

**Comment by elizabethandrews** [comment]
> > > While I understand the motivation here, I think that there isn't a good way about this.
> > > The __cplusplus macro is inconsistent between compilers that otherwise would be compatible (such as, we SHOULD be OK with a slightly older version of Clang where C++2a was 201911L).
> > > Additionally, IS there a problem if the host is a slightly newer/older version?
> > > Finally, this is going to be a maintenance headache (the rarity of is change, every 3 years, is going to result in us forgetting to update this) and massively error prone, and doesn't do what we need it to do.
> > > I just think this is a bad idea.
> > 
> > 
> > @erichkeane can you take a look at the latest patch. It has been modified to throw the error only when host version is C++11 (or less) and device version is C++14 (or greater)
> 
> This doesn't solve the problem actually. The issue is that MSVC until recently (and on hosts we want to support), made this number always 1997. It still supported a large amount of C++17 features at that point.

Hmm...How can we resolve this then? For MSVC, how do we detect if host and device C++ versions are incompatible if MSVC (and other compilers if any) doesn't maintain this consistently?

**Comment by erichkeane** [comment]
> 
> 
> > > > While I understand the motivation here, I think that there isn't a good way about this.
> > > > The __cplusplus macro is inconsistent between compilers that otherwise would be compatible (such as, we SHOULD be OK with a slightly older version of Clang where C++2a was 201911L).
> > > > Additionally, IS there a problem if the host is a slightly newer/older version?
> > > > Finally, this is going to be a maintenance headache (the rarity of is change, every 3 years, is going to result in us forgetting to update this) and massively error prone, and doesn't do what we need it to do.
> > > > I just think this is a bad idea.
> > > 
> > > 
> > > @erichkeane can you take a look at the latest patch. It has been modified to throw the error only when host version is C++11 (or less) and device version is C++14 (or greater)
> > 
> > 
> > This doesn't solve the problem actually. The issue is that MSVC until recently (and on hosts we want to support), made this number always 1997. It still supported a large amount of C++17 features at that point.
> 
> Hmm...How can we resolve this then? For MSVC, how do we detect if host and device C++ versions are incompatible if MSVC (and other compilers if any) doesn't maintain this consistently?

That is why I said initially (I think...) that there just isn't a good way to implement this.  Even the current implementation is going to just NOT WORK on any pre-2019 MSVC.  We could detect that and just bail (#if _MSVC && __cplusplus==1997xxxx), but I'm still unconvinced this problem is sufficiently motivated.

**Comment by rolandschulz** [comment]
> > > > While I understand the motivation here, I think that there isn't a good way about this.
> > > > The __cplusplus macro is inconsistent between compilers that otherwise would be compatible (such as, we SHOULD be OK with a slightly older version of Clang where C++2a was 201911L).
> > > > Additionally, IS there a problem if the host is a slightly newer/older version?
> > > > Finally, this is going to be a maintenance headache (the rarity of is change, every 3 years, is going to result in us forgetting to update this) and massively error prone, and doesn't do what we need it to do.
> > > > I just think this is a bad idea.
> > > 
> > > 
> > > @erichkeane can you take a look at the latest patch. It has been modified to throw the error only when host version is C++11 (or less) and device version is C++14 (or greater)
> > 
> > 
> > This doesn't solve the problem actually. The issue is that MSVC until recently (and on hosts we want to support), made this number always 1997. It still supported a large amount of C++17 features at that point.
> 
> Hmm...How can we resolve this then? For MSVC, how do we detect if host and device C++ versions are incompatible if MSVC (and other compilers if any) doesn't maintain this consistently?

In theory it is possible to use [`_MSVC_LANG` to check](https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/). But that is always at least C++14. Therefore if just want to check for >=C++14 it is sufficient to check for `__cplusplus > 201112L || defined(_MSC_VER)`. Given that MSVC doesn't support `/std:c++11` (or earlier) as version flag.

**Comment by elizabethandrews** [comment]
> __cplusplus > 201112L || defined(_MSC_VER)

Thanks for explanation. I think we need to check for >=C++14 when _generating_ the integration header (device) and <= C++11 when _using_ the integration header (host). For the former, @erichkeane suggested using LangOpts.CPlusPlus14. Don't know off the top of my head, how clang sets this. Might be interesting to see how MSVC is handled by clang for this LangOpt.

**Comment by elizabethandrews** [comment]
> > > > > While I understand the motivation here, I think that there isn't a good way about this.
> > > > > The __cplusplus macro is inconsistent between compilers that otherwise would be compatible (such as, we SHOULD be OK with a slightly older version of Clang where C++2a was 201911L).
> > > > > Additionally, IS there a problem if the host is a slightly newer/older version?
> > > > > Finally, this is going to be a maintenance headache (the rarity of is change, every 3 years, is going to result in us forgetting to update this) and massively error prone, and doesn't do what we need it to do.
> > > > > I just think this is a bad idea.
> > > > 
> > > > 
> > > > @erichkeane can you take a look at the latest patch. It has been modified to throw the error only when host version is C++11 (or less) and device version is C++14 (or greater)
> > > 
> > > 
> > > This doesn't solve the problem actually. The issue is that MSVC until recently (and on hosts we want to support), made this number always 1997. It still supported a large amount of C++17 features at that point.
> > 
> > 
> > Hmm...How can we resolve this then? For MSVC, how do we detect if host and device C++ versions are incompatible if MSVC (and other compilers if any) doesn't maintain this consistently?
> 
> In theory it is possible to use [`_MSVC_LANG` to check](https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/). But that is always at least C++14. Therefore if just want to check for >=C++14 it is sufficient to check for `__cplusplus > 201112L || defined(_MSC_VER)`. Given that MSVC doesn't support `/std:c++11` (or earlier) as version flag.

Thanks for explanation. I think we need to check for >=C++14 when _generating_ the integration header (device) and <= C++11 when _using_ the integration header (host). For the former, @erichkeane suggested using LangOpts.CPlusPlus14. Don't know off the top of my head, how clang sets this. Might be interesting to see how MSVC is handled by clang for this LangOpt.

**Comment by elizabethandrews** [comment]
> > > > > While I understand the motivation here, I think that there isn't a good way about this.
> > > > > The __cplusplus macro is inconsistent between compilers that otherwise would be compatible (such as, we SHOULD be OK with a slightly older version of Clang where C++2a was 201911L).
> > > > > Additionally, IS there a problem if the host is a slightly newer/older version?
> > > > > Finally, this is going to be a maintenance headache (the rarity of is change, every 3 years, is going to result in us forgetting to update this) and massively error prone, and doesn't do what we need it to do.
> > > > > I just think this is a bad idea.
> > > > 
> > > > 
> > > > @erichkeane can you take a look at the latest patch. It has been modified to throw the error only when host version is C++11 (or less) and device version is C++14 (or greater)
> > > 
> > > 
> > > This doesn't solve the problem actually. The issue is that MSVC until recently (and on hosts we want to support), made this number always 1997. It still supported a large amount of C++17 features at that point.
> > 
> > 
> > Hmm...How can we resolve this then? For MSVC, how do we detect if host and device C++ versions are incompatible if MSVC (and other compilers if any) doesn't maintain this consistently?
> 
> That is why I said initially (I think...) that there just isn't a good way to implement this. Even the current implementation is going to just NOT WORK on any pre-2019 MSVC. We could detect that and just bail (#if _MSVC && __cplusplus==1997xxxx), but I'm still unconvinced this problem is sufficiently motivated.

@JoeyGIntel What is the motivation for this patch? What issues are we seeing?

**Comment by erichkeane** [comment]
> 
> 
> > > > > While I understand the motivation here, I think that there isn't a good way about this.
> > > > > The __cplusplus macro is inconsistent between compilers that otherwise would be compatible (such as, we SHOULD be OK with a slightly older version of Clang where C++2a was 201911L).
> > > > > Additionally, IS there a problem if the host is a slightly newer/older version?
> > > > > Finally, this is going to be a maintenance headache (the rarity of is change, every 3 years, is going to result in us forgetting to update this) and massively error prone, and doesn't do what we need it to do.
> > > > > I just think this is a bad idea.
> > > > 
> > > > 
> > > > @erichkeane can you take a look at the latest patch. It has been modified to throw the error only when host version is C++11 (or less) and device version is C++14 (or greater)
> > > 
> > > 
> > > This doesn't solve the problem actually. The issue is that MSVC until recently (and on hosts we want to support), made this number always 1997. It still supported a large amount of C++17 features at that point.
> > 
> > 
> > Hmm...How can we resolve this then? For MSVC, how do we detect if host and device C++ versions are incompatible if MSVC (and other compilers if any) doesn't maintain this consistently?
> 
> In theory it is possible to use [`_MSVC_LANG` to check](https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/). But that is always at least C++14. Therefore if just want to check for >=C++14 it is sufficient to check for `__cplusplus > 201112L || defined(_MSC_VER)`. Given that MSVC doesn't support `/std:c++11` (or earlier) as version flag.

This actually doesn't work for the compilers from microsoft BEFORE the switch to correctly reporting __cplusplus, right?  The ones I'm concerned about are when __cplusplus reported 199711L despite having sufficient C++14 or C++17 support.  However, properly using the _MSC_VER value would require figuring out the first version of MSVC that we believe matches the required list of supported features.

**Comment by erichkeane** [comment]
> 
> 
> > > > > > While I understand the motivation here, I think that there isn't a good way about this.
> > > > > > The __cplusplus macro is inconsistent between compilers that otherwise would be compatible (such as, we SHOULD be OK with a slightly older version of Clang where C++2a was 201911L).
> > > > > > Additionally, IS there a problem if the host is a slightly newer/older version?
> > > > > > Finally, this is going to be a maintenance headache (the rarity of is change, every 3 years, is going to result in us forgetting to update this) and massively error prone, and doesn't do what we need it to do.
> > > > > > I just think this is a bad idea.
> > > > > 
> > > > > 
> > > > > @erichkeane can you take a look at the latest patch. It has been modified to throw the error only when host version is C++11 (or less) and device version is C++14 (or greater)
> > > > 
> > > > 
> > > > This doesn't solve the problem actually. The issue is that MSVC until recently (and on hosts we want to support), made this number always 1997. It still supported a large amount of C++17 features at that point.
> > > 
> > > 
> > > Hmm...How can we resolve this then? For MSVC, how do we detect if host and device C++ versions are incompatible if MSVC (and other compilers if any) doesn't maintain this consistently?
> > 
> > 
> > In theory it is possible to use [`_MSVC_LANG` to check](https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/). But that is always at least C++14. Therefore if just want to check for >=C++14 it is sufficient to check for `__cplusplus > 201112L || defined(_MSC_VER)`. Given that MSVC doesn't support `/std:c++11` (or earlier) as version flag.
> 
> Thanks for explanation. I think we need to check for >=C++14 when _generating_ the integration header (device) and <= C++11 when _using_ the integration header (host). For the former, @erichkeane suggested using LangOpts.CPlusPlus14. Don't know off the top of my head, how clang sets this. Might be interesting to see how MSVC is handled by clang for this LangOpt.

The LangOpts c++ versions are all set if AT LEAST that set is supported.  For example, CPlusPlus17 ALSO sets CPlusPlus14, CPlusPlus11 and CPlusPlus.  So checking CPlusPlus14 is sufficient to check for AT LEAST C++14.
